File: backend/models/session.py
import enum
import uuid
from datetime import datetime

from sqlalchemy import DateTime, Enum, ForeignKey, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from database.base import Base


class SessionStatus(str, enum.Enum):
    SCHEDULED = "SCHEDULED"
    COMPLETED = "COMPLETED"
    CANCELLED = "CANCELLED"
    RESCHEDULED = "RESCHEDULED"


class SessionMeetingType(str, enum.Enum):
    GOOGLE_MEET = "GOOGLE_MEET"
    ZOOM = "ZOOM"
    WHATSAPP = "WHATSAPP"
    PHONE = "PHONE"
    IN_PERSON = "IN_PERSON"


class SessionAttendanceStatus(str, enum.Enum):
    PRESENT = "PRESENT"
    BOTH_PRESENT = "BOTH_PRESENT"
    CLIENT_NO_SHOW = "CLIENT_NO_SHOW"
    TRAINER_NO_SHOW = "TRAINER_NO_SHOW"
    LATE = "LATE"
    RESCHEDULED = "RESCHEDULED"


class Session(Base):
    """A scheduled coaching interaction between a Client and a TrainerProfile.

    Version-1 scheduling philosophy: sessions are never auto-generated from
    subscriptions. A Client subscribes, a Trainer is assigned, the Trainer
    agrees on a schedule with the client out of band, and the Trainer (or a
    SUPER_ADMIN acting as an override/escalation role) manually creates each
    Session row. Bulk creation (e.g. recurring weekly slots) is a Task-17.2
    concern and is intentionally not implemented here.

    Business rules enforced in later tasks (not this one):
      1. Client must have an eligible subscription.
      2. Trainer must be assigned to the client.
      3. Sessions cannot be scheduled in the past.
      4. Trainers can only manage their assigned clients.
      5. Clients can only view their own sessions.
      6. SUPER_ADMIN has full access.
      7. Session overlap validation will be implemented later.
      8. Remaining subscription session validation will be implemented later.
    """

    __tablename__ = "sessions"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    client_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("clients.id"), nullable=False
    )
    trainer_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("trainer_profiles.id"), nullable=False
    )

    scheduled_start: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )
    scheduled_end: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )
    duration_minutes: Mapped[int] = mapped_column(
        Integer, nullable=False, default=60
    )

    status: Mapped[SessionStatus] = mapped_column(
        Enum(SessionStatus, name="session_status"), nullable=False
    )

    meeting_type: Mapped[SessionMeetingType] = mapped_column(
        Enum(SessionMeetingType, name="session_meeting_type"), nullable=False
    )
    # Annotation kept non-Optional (with nullable=True set explicitly) because
    # SQLAlchemy 2.0.36's Mapped[X | None] resolution crashes on Python 3.14;
    # the column is nullable at the DB/ORM level regardless.
    meeting_link: Mapped[str] = mapped_column(String(500), nullable=True)

    trainer_notes: Mapped[str] = mapped_column(Text, nullable=True)
    trainer_feedback: Mapped[str] = mapped_column(Text, nullable=True)
    homework: Mapped[str] = mapped_column(Text, nullable=True)
    next_session_focus: Mapped[str] = mapped_column(Text, nullable=True)

    # Nullable: attendance is only known once the session has taken place,
    # so a freshly SCHEDULED session has no attendance_status yet.
    attendance_status: Mapped[SessionAttendanceStatus] = mapped_column(
        Enum(SessionAttendanceStatus, name="session_attendance_status"),
        nullable=True,
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    client: Mapped["Client"] = relationship(  # noqa: F821
        "Client", foreign_keys=[client_id], back_populates="sessions"
    )
    trainer: Mapped["TrainerProfile"] = relationship(  # noqa: F821
        "TrainerProfile", foreign_keys=[trainer_id], back_populates="sessions"
    )

    def __repr__(self) -> str:
        return (
            f"Session(id={self.id!r}, client_id={self.client_id!r}, "
            f"trainer_id={self.trainer_id!r}, status={self.status!r})"
        )

File: frontend/src/types/session.ts
import type { PaginatedResponse } from "@/types/dashboard"

export type SessionStatus = "SCHEDULED" | "COMPLETED" | "CANCELLED" | "RESCHEDULED"

export type SessionMeetingType = "GOOGLE_MEET" | "ZOOM" | "WHATSAPP" | "PHONE" | "IN_PERSON"

export type SessionAttendanceStatus =
  | "PRESENT"
  | "BOTH_PRESENT"
  | "CLIENT_NO_SHOW"
  | "TRAINER_NO_SHOW"
  | "LATE"
  | "RESCHEDULED"

export type Weekday = "MONDAY" | "TUESDAY" | "WEDNESDAY" | "THURSDAY" | "FRIDAY" | "SATURDAY" | "SUNDAY"

export interface Session {
  id: string
  client_id: string
  trainer_id: string
  scheduled_start: string
  scheduled_end: string
  duration_minutes: number
  status: SessionStatus
  meeting_type: SessionMeetingType
  meeting_link: string | null
  trainer_notes: string | null
  trainer_feedback: string | null
  homework: string | null
  next_session_focus: string | null
  attendance_status: SessionAttendanceStatus | null
  created_at: string
  updated_at: string
}

export type PaginatedSessions = PaginatedResponse<Session>

export interface SessionCreateInput {
  client_id: string
  trainer_id?: string | null
  scheduled_start: string
  duration_minutes: number
  meeting_type: SessionMeetingType
  meeting_link?: string | null
  trainer_notes?: string | null
}

export interface SessionBulkCreateInput {
  client_id: string
  trainer_id?: string | null
  start_date: string
  end_date: string
  days: Weekday[]
  start_time: string
  duration_minutes: number
  meeting_type: SessionMeetingType
}

export interface SessionBulkCreateResult {
  sessions_created: number
  sessions_skipped: number
  skipped_reasons: string[]
}

/** Only scheduled_start, meeting_type, meeting_link, and status are editable
 * via `PATCH /sessions/{id}` per Task 22.6 - attendance and notes go through
 * their own dedicated endpoints/inputs below. */
export interface SessionUpdateInput {
  scheduled_start?: string
  meeting_type?: SessionMeetingType
  meeting_link?: string | null
  status?: SessionStatus
}

export interface SessionNotesUpdateInput {
  trainer_notes?: string | null
  trainer_feedback?: string | null
  homework?: string | null
  next_session_focus?: string | null
}

export interface SessionAttendanceUpdateInput {
  attendance_status: SessionAttendanceStatus
}

/** Narrower view used by the Client "My Sessions" screens - deliberately
 * omits trainer_notes, trainer_feedback, and next_session_focus, which
 * clients must never see (Task 22.6 requirement 8). */
export type ClientSessionView = Omit<Session, "trainer_notes" | "trainer_feedback" | "next_session_focus">
